### ISSUE TITLE
Feature request: drop 'private', 'public', 'public export'. Introduce 'open', 'close' 

### ISSUE THREAD

**Issue by nicolabotta** [idris-lang/Idris-dev#3198]
Consider the program
```idris
> module EliminateExportQualifiers
 
> import Syntax.PreorderReasoning

> import NonNegRational
> import NonNegRationalBasicOperations
> import NonNegRationalBasicProperties
> import Fraction
> import FractionNormal
> import NatPositive
> import ListProperties
 
> %default total
> %access public export

> factor : {A : Type} -> List A -> NonNegRational
> factor Nil = 1
> factor (a :: as) = fromFraction (1, Element (S (length as)) MkPositive)

> -- %freeze fromFraction

> factorLemma : {A, B, C : Type} -> 
>               (as : List A) -> (f : A -> B) -> (g : A -> C) ->
>               factor (map f as) = factor (map g as)
> factorLemma  Nil f g = Refl
> factorLemma (a :: as) f g = 
>     ( factor (map f (a :: as)) )
>   ={ Refl }=
>     ( factor (f a :: map f as) )
>   ={ Refl }=
>     ( fromFraction (1, Element (S (length (map f as))) MkPositive) )
>   ={ cong {f = \ ZUZU => fromFraction (1, Element (S ZUZU) MkPositive)} (lengthLemma as f g) }=
>     ( fromFraction (1, Element (S (length (map g as))) MkPositive) )  
>   ={ Refl }=
>     ( factor (g a :: map g as) )
>   ={ Refl }=
>     ( factor (map g (a :: as)) )
>   QED
```
On a fresh installation of `https://github.com/nicolabotta/SeqDecProbs`, you should be able to do
```
$ cd SeqDecProbs-master/frameworks/14-/
$ time idris +RTS -K32000000 -RTS -p contrib -V --check EliminateExportQualifiers.lidr
```
and get something like
```
Type checking ./Sigma.lidr
Type checking ./PairsOperations.lidr
Type checking ./Finite.lidr
Type checking ./UnitProperties.lidr
Type checking ./VoidProperties.lidr
...
Type checking ./NonNegRationalPredicates.lidr
Type checking ./NonNegRationalBasicProperties.lidr
Type checking ./EliminateExportQualifiers.lidr

real	30m16.885s
user	30m7.072s
sys	0m10.392s
```
It is easy to see that type cheking `EliminateExportQualifiers.lidr` alone takes about 4.5 minutes out of the 30 minutes that it takes to type check the file together with its dependencies:
```
$ rm EliminateExportQualifiers.ibc
$ time idris +RTS -K32000000 -RTS -p contrib -V --check EliminateExportQualifiers.lidr
Type checking ./EliminateExportQualifiers.lidr

real	4m30.878s
user	4m28.200s
sys	0m2.616s
```
**The interesting observation is that uncommeting line 20** 
```
> -- %freeze fromFraction
```
**of `EliminateExportQualifiers.lidr` reduces the type checking time from 4.5 minutes to about 20 seconds!** 

I believe that the above example allows us to draw the following conclusions:

1) There is significant space for improving the type checking process. In particular, the type checker seems to be overly eager after expanding available (unfrozen, publicly exported) expressions even when this is not necessary for type checking to succeed.

2) In all but the most rare cases, there seems to be little use for the `private`, `public` and `public export` qualifiers: in the concrete example, for instance, it would make no sense to declare `NonNegRationalBasicOperations.fromFraction` as `private` or even `public`. This is because there are function (e.g., in `NonNegRationalBasicOperations.lidr`) that (in contrast to `factorLemma`) do need to access the definition of `fromFraction`. In fact, it seems that, except perhaps for application-level files, the only meaningful accessibility default is `public export`. A cursory look at the prelude confirms that `%access public export` is ubiquitous.

3) In contrast, it would be extremely useful (and, I argue, necessary for sizeable applications) to be able to `open` a function's implementation for a well defined scope.

_Thus, the feature request is to drop the `private`, `public` and `public export` from Idris and make all declarations per default publicly exported._

_At the same time, please support `open` and `close` (`thaw` and `freeze`) clauses to fine grain control the access to function implementations, data constructor, etc. and thus prevent the type checker to waste huge amounts of computational resources._

**Comment by jfdm** [comment]
Just to comment on point (2), the other points I am not in a position to provide an informed comment.

I use the accessibility modifiers all the time in my programs. I use them to provide a succinct description of the public API for a module. This will mean that some things in the same module might be: private---because I don't want to expose them; public export---because I want to expose them completely constructors and all; and export---as their descriptions are more useful than implementations. This comes from the [descriptions of the modifiers in the Idris documentation](http://idris.readthedocs.io/en/latest/tutorial/modules.html).

Your last comment on open and close, are exactly how I use Idris' current accessibility modifiers.

AFAIK, the prelude is default `public export` as a fine-grained revisiting of the accessibility for each construct in each module in the prelude has yet to occur. The default has been `public export` as 'verbose' API is not necessarily unwanted.

Maybe we are not clear enough on how to use the modifiers in Idris code. This is something that should be revised, I know I have a PR that needs finishing to document how I use the accessibility modifiers.

As for `thaw` and `freeze`, I am not familiar enough with them to comment.

**Comment by ahmadsalim** [comment]
If I understand some things correctly, then @nicolabotta proposes something which I think is complementary in some ways to the `public export`, etc. modifiers which is to allow to selectively importing definitions depending on need. 

I would guess the issue is that sometimes you want to expose definitions to other modules from a particular module A, but not all modules that import A necessarily care about the said definitions.

**Comment by ziman** [comment]
To put it another way, I think that Nicola proposes declaring opaqueness at the point of usage, not at the point of definition.

**Comment by edwinb** [comment]
I think this would be a workaround, not a fix. There is a better way to fix to amount of resources being used (which is to use weak head reduction in the type checker rather than always normalising), but I'm juggling far too many things to be able to do it at the moment, I'm afraid. 

This *will* be done before version 1.0 - I consider it a release blocker. In fact, I think it's what causes the behaviour in #172 and it's fairly simple to write a small definition that really hurts performance.

I'm certainly not going to drop the export modifiers for version 1.0, though, so I'm going to close this. If someone wants to add a feature that allows finer grained control themselves, that might be a different matter - please carry on discussing here!

**Comment by edwinb** [comment]
For the record, here's a program that exhibits the underlying problem:

    data IsSuc : Nat -> Type where
         YesItIs : IsSuc (S x)
   
    ouch : IsSuc (100 * 100)
    ouch = YesItIs

I think I'll post this as an independent issue actually, just to make sure it's addressed.

**Comment by edwinb** [comment]
Also, just to throw this in too, I and a couple of others are currently working on how to use interfaces as the preferred way of abstracting things. There's still a bit to do, but it's looking like it'll do the job pretty well (and it appears that interfaces in their current form are remarkably close to ML modules...).

More on this later, though, when we've actually worked it out...

**Comment by nicolabotta** [comment]
> I use the accessibility modifiers all the time in my programs. I use them to provide a succinct description of the public API for a module. This will mean that some things in the same module might be: private---because I don't want to expose them; public export---because I want to expose them completely constructors and all; and export---as their descriptions are more useful than implementations. This comes from the descriptions of the modifiers in the Idris documentation.

The problem is that what you want (or do not want) has very little bearing on what a client of your module actually needs (or does not need). It is commonly the case that the implementation of an `A.foo` function is required to be accessible in a `B` module that imports `A` and unaccessible in a `C` module that also imports `A`. This is typically the case when `B` contains proofs of properties of `A.foo` and `C` only applies `A.foo`.

The point that I am trying to make is that you cannot, in general, know whether a module that imports your module needs access to a function definition or not. This is particularly true in a dependently typed language: someone (perhaps you) might need to prove some properties of one of your functions. In this case, he will most likely need access to the implementation. If you have declared that function to be `private` or `public`, he will need to duplicate it and make it `public export` available. This leads to duplication of code, inefficiencies, poor maintainability. Conversely, if you have declared a function to be `public export`, some of your clients might experience unacceptable type checking times (as in the example above) because the type checker will expand your definitions even in situations in which this is not needed and, in fact, contraproductive.

I am not in principle against accessibility annotations. These could be very useful to signal a developer's intention. But I very much doubt that it makes sense (again, in general) to strictly prevent clients from being able to access a function definition. At the same time, it appears to be necessary for clients to be able to restrict access to certain features of imported modules, e.g., as shown above, to avoid unacceptable type checking inefficiencies.

In a nutshell: accessibility should be enforced at the point at which it is precisely know which features of a given module needs to be accessible and which can (or need to) be hidden. This is in the client code where those features are actually used.

**Comment by nicolabotta** [comment]
> I would guess the issue is that sometimes you want to expose definitions to other modules from a particular module A, but not all modules that import A necessarily care about the said definitions.

Yes, developers can use access qualifiers to indicate their intents. But clients must be allowed to override such recommendations. In dependently typed languages, client code might contain proofs of properties of imported features. Thus, only clients can know whether a specific imported feature must be accessible or not.

**Comment by nicolabotta** [comment]
> Also, just to throw this in too, I and a couple of others are currently working on how to use interfaces as the preferred way of abstracting things. There's still a bit to do, but it's looking like it'll do the job pretty well (and it appears that interfaces in their current form are remarkably close to ML modules...).

> More on this later, though, when we've actually worked it out...

Sounds very exciting ... perhaps ... we will be able to write something like
```
Foo : Type
Foo implements Num
```
to require `Foo` t be a type that implements `Num`? That would be really cool ...

**Comment by nicolabotta** [comment]
> To put it another way, I think that Nicola proposes declaring opaqueness at the point of usage, not at the point of definition.

Exactly. To stay with the above example, you can add 
```
> monotoneSum : {A : Type} ->
>               (f : A -> NonNegRational) -> (g : A -> NonNegRational) ->
>               (p : (a : A) -> f a `LTE` g a) ->
>               (as : List A) ->
>               sum (map f as) `LTE` sum (map g as) 
> monotoneSum f g p Nil = LTEZero
> monotoneSum f g p (a :: as) = s5 where
>   s1 : sum (map f (a :: as)) = f a + sum (map f as)
>   s1 = Refl
>   s2 : sum (map g (a :: as)) = g a + sum (map g as)
>   s2 = Refl
>   s3 : f a `LTE` g a
>   s3 = p a
>   s4 : sum (map f as) `LTE` sum (map g as)
>   s4 = monotoneSum f g p as
>   s5 : sum (map f (a :: as)) `LTE` sum (map g (a :: as))
>   s5 = NonNegRationalLTEProperties.monotonePlusLTE s3 s4
```
to `EliminateExportQualifiers.lidr`. This will type check fine (albeit in about 6 minutes!) **if** 
```
> %freeze fromFraction
```
is commented out. **Otherwise**, it will fail because the implementation of `fromFraction` is needed to type check `monotoneSum`. Thus, under the current type checking rules, we need to be able to `freeze` and `unfreeze` specific names for single functions. This is not possible with Idris 0.11.2. 

As shown in #3197, because of this issue Idris does not scale up very well and it is quite difficult to build sizeable Idris applications that type check in an acceptable time.

**Comment by nicolabotta** [comment]
> @nicolabotta I have removed some of the boldness in your comment since it was a bit hard to read. I hope it was OK.

Thanks Ahmad! Your formatting/highlightning looks much better than mine! Best, Nicola

**Comment by nicolabotta** [comment]
> @nicolabotta I think it is **sometimes** wise to disallow overriding such modifiers, i.e. if you as a library writer want to provide an interfaces and ensure that you are able to change the interface without breaking users code. One could see this as a kind of rely-guarantee reasoning, where one should only expect to rely on definitions of functions which have public export.

I do not mind sometimes allowing disallowing overriding such modifiers. But let's keep things in a perspective: the normal case still seems to be: **in general allow overriding**!
